ttnn/cpp/ttnn/tensor/vector_base.cpp

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "vector_base.hpp"
+#include <stdexcept>
+#include "fmt/color.h"
+#include "tt_metal/common/assert.hpp"
+
+namespace tt::tt_metal {
+
+namespace {
+
+int32_t normalized_index(int32_t index, size_t original_size, size_t container_size) {
+    int32_t orig_size = static_cast<int32_t>(original_size);
+    int32_t full_size = static_cast<int32_t>(container_size);
+
+    int fixed_index = index;
+    if (fixed_index < 0) {
+        fixed_index += full_size;
+    } else {
+        fixed_index += full_size - orig_size;
+    }
+
+    if (fixed_index < 0 || fixed_index >= full_size) {
+        TT_THROW("VectorBase[] index out of range. {} not in [{}, {})", index, -full_size, full_size);
+    }
+
+    return fixed_index;
+}
+}
+
+void VectorBase::init() {
+    m_original_size = m_value.size();
+    const size_t min_internal_size = 4;
+
+    if(m_original_size < min_internal_size) {
+        std::vector<uint32_t> ones(min_internal_size - m_original_size, 1);
+        m_value.insert(m_value.begin(), ones.begin(), ones.end());
+    }
+}
+
+size_t VectorBase::size() const {
+    return m_original_size;
+}
+
+std::span<const uint32_t> VectorBase::view() const {
+    return std::span<const uint32_t>(cbegin(), cend());
+}
+
+std::vector<uint32_t> VectorBase::as_vector() const {
```

Comment:
I believe as_vector() shouldn't be needed, I killed it SmallVector PR

Reply:
Awesome!